src/enumerable.cr

```diff
@@ -1747,11 +1747,21 @@ module Enumerable(T)
   # ```
   # ["a", "A", "b", "B"].tally_by(&.downcase) # => {"a" => 2, "b" => 2}
   # ```
-  def tally_by(& : T -> U) : Hash(U, Int32) forall U
-    each_with_object(Hash(U, Int32).new) do |item, hash|
+  #
+  # If an optional `hash` is given, the number of occurrences is added to each value
+  # in the `hash`, and the `hash` is returned.
+  #
+  # ```
+  # hash = {} of Char => Int32
+  # words = ["crystal", "ruby"]
+  # words.each { |word| word.chars.tally(hash) }
+  # hash
+  # => {'c' => 1, 'r' => 2, 'y' => 2, 's' => 1, 't' => 1, 'a' => 1, 'l' => 1, 'u' => 1, 'b' => 1}
+  # ```
+  def tally_by(hash = Hash(U, Int32).new, & : T -> U) : Hash(U, Int32) forall U
```

Comment:
Is that because someone could pass in a Hash that doesn't conform to `Hash(U, Int32)`?  If so, could this also be changed to the following?

```
  def tally_by(hash = Hash(U, Int32).new : Hash(U, Int32), & : T -> U) : Hash(U, Int32) forall U
```

Reply:
I am just curious how it can be a different type.

For example if you try

```
hash = {} of String => Int32
words = ["Crystal", "Ruby"]
words.each { |word| word.chars.tally_by(hash, &.downcase) }
```

The compiler will fail with:

```
Error: no overload matches 'Hash(String, Int32)#[]=' with types Char, Int32
```

Reply:
There can be legitimate values that do not conform to `Hash(U, Int32)` on different levels. The key type could be more broad than the return type of the block (`Hash(String | Char, Int32)`). The value type could probably be a different number type (`Hash(Char, Int8)`). The entire thing doesn't even need to be a `Hash` at all, it just needs to expose a hash-like interface.

Reply:
Rever back to two separate methods.
But I didn't get about "different type than `Hash(U, Int32)`".

Can someone point me how to change the `tally_by(hash : Hash(U, Int32), & : T -> U) : Hash(U, Int32) forall U` method to meet this requirement? Thanks.

Reply:
tally_by(hash)